README.md
colorify [![CircleCI](https://circleci.com/gh/ahmedkamals/colorify.svg?style=svg)](https://circleci.com/gh/ahmedkamals/colorify "Build Status")
==========

[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](LICENSE  "License")
[![release](https://img.shields.io/github/release/ahmedkamals/colorify.svg?style=flat-square)](https://github.com/ahmedkamals/colorify/releases/latest "Release")
[![Travis CI](https://travis-ci.org/ahmedkamals/colorify.svg)](https://travis-ci.org/ahmedkamals/colorify "Cross Build Status [Linux, OSx]") 
[![codecov](https://codecov.io/gh/ahmedkamals/colorify/branch/master/graph/badge.svg)](https://codecov.io/gh/ahmedkamals/colorify "Code Coverage")
[![Coverage Status](https://coveralls.io/repos/github/ahmedkamals/colorify/badge.svg?branch=master)](https://coveralls.io/github/ahmedkamals/colorify?branch=master  "Code Coverage")
[![GolangCI](https://golangci.com/badges/github.com/ahmedkamals/colorify.svg?style=flat-square)](https://golangci.com/r/github.com/ahmedkamals/colorify "Code Coverage")
[![Go Report Card](https://goreportcard.com/badge/github.com/ahmedkamals/colorify)](https://goreportcard.com/report/github.com/ahmedkamals/colorify  "Go Report Card")
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/c282df1ff33c43ddb5da1d7fe4e85674)](https://www.codacy.com/app/ahmedkamals/colorify?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ahmedkamals/colorify&amp;utm_campaign=Badge_Grade "Code Quality")
[![Go Walker](http://gowalker.org/api/v1/badge)](https://gowalker.org/github.com/ahmedkamals/colorify "Documentation")
[![GoDoc](https://godoc.org/github.com/ahmedkamals/colorify?status.svg)](https://godoc.org/github.com/ahmedkamals/colorify "API Documentation")
[![Docker Pulls](https://img.shields.io/docker/pulls/ahmedkamal/colorify.svg?maxAge=604800)](https://hub.docker.com/r/ahmedkamal/colorify/ "Docker Pulls")
[![DepShield Badge](https://depshield.sonatype.org/badges/ahmedkamals/colorify/depshield.svg)](https://depshield.github.io "DepShield")
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fahmedkamals%2Fcolorify.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fahmedkamals%2Fcolorify?ref=badge_shield "Dependencies")
[![Join the chat at https://gitter.im/ahmedkamals/colorify](https://badges.gitter.im/ahmedkamals/colorify.svg)](https://gitter.im/ahmedkamals/colorify?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge "Let's discuss")

```bash

 _____       _            _  __       
/  __ \     | |          (_)/ _|      
| /  \/ ___ | | ___  _ __ _| |_ _   _ 
| |    / _ \| |/ _ \| '__| |  _| | | |
| \__/\ (_) | | (_) | |  | | | | |_| |
 \____/\___/|_|\___/|_|  |_|_|  \__, |
                                 __/ |
                                |___/ 

```

is a library that helps applying RGB colors based on [24 bit - ANSI escape sequences][1] for console output.

Table of Contents
-----------------

*   [🏎️ Getting Started](#getting-started)

    *   [Prerequisites](#prerequisites)
    *   [Example](#example)

*   [🕸️ Tests](#tests)

    *   [Benchmarks](#benchmarks)

*   [🤝 Contribution](#-contribution)

*   [👨‍💻 Author](#-author)

*   [🆓 License](#-license)

🏎️ Getting Started
------------------

### Prerequisites

*   [Golang 1.13 or later][2].

### Example

```go
    colorize := colorify.NewColorable(os.Stdout)
	style := colorify.Style{
		Foreground: &colorify.ColorValue{
			Red:   88,
			Green: 188,
			Blue:  88,
		},
		Background: &colorify.ColorValue{
			Red:   188,
			Green: 88,
			Blue:  8,
		},
		Font: []colorify.FontEffect{
			colorify.Bold,
			colorify.Italic,
			colorify.Underline,
			colorify.CrossedOut,
		},
	}
	println(colorize.Wrap("Hello styled", style))
	println(colorize.Black("Text in black!"))
	println(colorize.Blue("Deep clue C!"))
	println(colorize.Cyan("Hello cyan!"))
	println(colorize.Gray("Gray logged text!"))
	println(colorize.Green("50 shades of Green!"))
	println(colorize.Magenta("Magenta!"))
	println(colorize.Red("The thin Red light!"))
	println(colorize.Orange("Orange is the new black!"))

	colorize.Set(style)
	print("Styled output!")
	colorize.Reset()
	println("\n\nSample Colors\n==============\n")
	println(colorize.Sample())
```

![Sample output](https://github.com/ahmedkamals/colorify/raw/master/assets/img/sample.png)

🕸️ Tests
--------
    
```bash
$ make test
```

### Benchmarks

![Benchmarks](https://github.com/ahmedkamals/colorify/raw/master/assets/img/bench.png)

🤝 Contribution
---------------

Please refer to [CONTRIBUTING.md](https://github.com/ahmedkamals/colorify/blob/master/CONTRIBUTING.md) file.

👨‍💻 Author
-----------

[ahmedkamals][3]

🆓 LICENSE
----------

Colorify is released under MIT license, refer [LICENSE](https://github.com/ahmedkamals/colorify/blob/master/LICENSE) file.  
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fahmedkamals%2Fcolorify.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fahmedkamals%2Fcolorify?ref=badge_large)


Enjoy!
[![Analytics](http://www.google-analytics.com/__utm.gif?utmwv=4&utmn=869876874&utmac=UA-136526477-1&utmcs=ISO-8859-1&utmhn=github.com&utmdt=colorify&utmcn=1&utmr=0&utmp=/ahmedkamals/colorify?utm_source=www.github.com&utm_campaign=colorify&utm_term=colorify&utm_content=colorify&utm_medium=repository&utmac=UA-136526477-1)]()

[1]: https://en.wikipedia.org/wiki/ANSI_escape_code#24-bit "Ansi Escape Sequenece"
[2]: https://golang.org/dl/ "Download Golang"
[3]: https://github.com/ahmedkamals "Author"


CHANGELOG.md
<!-- BEGIN MUNGE: GENERATED_TOC -->

-   [v0.0.1](#v001)

<!-- END MUNGE: GENERATED_TOC -->

<!-- NEW RELEASE NOTES ENTRY -->

# Versions

## v0.0.1

-   First version of colorify.


colorify_test.go
package colorify

import (
	"fmt"
	"github.com/stretchr/testify/assert"
	"io/ioutil"
	"os"
	"testing"
)

func TestWrap(t *testing.T) {
	testCases := []struct {
		id           string
		input        string
		appliedStyle Style
		expected     string
	}{
		{
			id:    "Should output in black color.",
			input: "Output this in black",
			appliedStyle: Style{
				Foreground: &ColorValue{
					Red:   0,
					Green: 0,
					Blue:  0,
				},
			},
			expected: "\x1b[38;2;0;0;0mOutput this in black\x1b[0m",
		},
		{
			id:    "Should output in bold Red color.",
			input: "Output this in bold Red",
			appliedStyle: Style{
				Foreground: &ColorValue{
					Red:   255,
					Green: 0,
					Blue:  0,
				},
				Font: []FontEffect{Bold},
			},
			expected: "\x1b[38;2;255;0;0;1mOutput this in bold Red\x1b[0m",
		},
		{
			id:    "Should output in bold italic Green color.",
			input: "Output this in bold italic Green",
			appliedStyle: Style{
				Foreground: &ColorValue{
					Red:   0,
					Green: 255,
					Blue:  0,
				},
				Font: []FontEffect{Bold, Italic},
			},
			expected: "\x1b[38;2;0;255;0;1;3mOutput this in bold italic Green\x1b[0m",
		},
	}

	colorize := NewColorable(os.Stdout)
	for _, testCase := range testCases {
		t.Run(testCase.id, func(t *testing.T) {
			assert.Equal(t, fmt.Sprintf("%q", testCase.expected), fmt.Sprintf("%q", colorize.Wrap(testCase.input, testCase.appliedStyle)))
		})
	}
}

func TestSet(t *testing.T) {
	testCases := []struct {
		id           string
		input        string
		appliedStyle Style
		expected     string
	}{
		{
			id:    "Should output in black color.",
			input: "Output this in black",
			appliedStyle: Style{
				Foreground: &ColorValue{
					Red:   0,
					Green: 0,
					Blue:  0,
				},
			},
			expected: "\x1b[38;2;0;0;0mOutput this in black\x1b[0m",
		},
	}

	for _, testCase := range testCases {
		t.Run(testCase.id, func(t *testing.T) {
			output := captureOutput(func(output *os.File) {
				colorize := NewColorable(output)
				colorize.Set(testCase.appliedStyle)
				fmt.Print(testCase.input)
				colorize.Reset()
			})

			assert.Equal(t, fmt.Sprintf("%q", testCase.expected), fmt.Sprintf("%q", output))
		})
	}
}

func TestReset(t *testing.T) {
	testCases := []struct {
		id           string
		input        string
		appliedStyle Style
		expected     string
	}{
		{
			id:    "Should output in normal color.",
			input: "Output this normally.",
			appliedStyle: Style{
				Foreground: &ColorValue{
					Red:   0,
					Green: 0,
					Blue:  0,
				},
			},
			expected: "\x1b[38;2;0;0;0m\x1b[0mOutput this normally.",
		},
	}

	for _, testCase := range testCases {
		t.Run(testCase.id, func(t *testing.T) {
			output := captureOutput(func(output *os.File) {
				colorize := NewColorable(output)
				colorize.Set(testCase.appliedStyle)
				colorize.Reset()
				fmt.Print(testCase.input)
			})

			assert.Equal(t, fmt.Sprintf("%q", testCase.expected), fmt.Sprintf("%q", output))
		})
	}
}

func TestDirectColors(t *testing.T) {
	testCases := []struct {
		id           string
		input        string
		appliedStyle func(*Colorable, string) string
		expected     string
	}{
		{
			id:    "Should output in black color.",
			input: "Output this in black color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Black(str)
			},
			expected: "\x1b[38;2;0;0;0mOutput this in black color.\x1b[0m",
		},
		{
			id:    "Should output in blue color.",
			input: "Output this in blue color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Blue(str)
			},
			expected: "\x1b[38;2;0;0;255mOutput this in blue color.\x1b[0m",
		},
		{
			id:    "Should output in cyan color.",
			input: "Output this in cyan color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Cyan(str)
			},
			expected: "\x1b[38;2;0;255;255mOutput this in cyan color.\x1b[0m",
		},
		{
			id:    "Should output in gray color.",
			input: "Output this in gray color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Gray(str)
			},
			expected: "\x1b[38;2;128;128;128mOutput this in gray color.\x1b[0m",
		},
		{
			id:    "Should output in green color.",
			input: "Output this in green color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Green(str)
			},
			expected: "\x1b[38;2;0;255;0mOutput this in green color.\x1b[0m",
		},
		{
			id:    "Should output in magenta color.",
			input: "Output this in magenta color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Magenta(str)
			},
			expected: "\x1b[38;2;255;0;255mOutput this in magenta color.\x1b[0m",
		},
		{
			id:    "Should output in orange color.",
			input: "Output this in orange color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Orange(str)
			},
			expected: "\x1b[38;2;255;165;0mOutput this in orange color.\x1b[0m",
		},
		{
			id:    "Should output in red color.",
			input: "Output this in red color.",
			appliedStyle: func(colorable *Colorable, str string) string {
				return colorable.Red(str)
			},
			expected: "\x1b[38;2;255;0;0mOutput this in red color.\x1b[0m",
		},
	}

	for _, testCase := range testCases {
		t.Run(testCase.id, func(t *testing.T) {
			colorize := NewColorable(os.Stdout)
			output := testCase.appliedStyle(colorize, testCase.input)

			assert.Equal(t, fmt.Sprintf("%q", testCase.expected), fmt.Sprintf("%q", output))
		})
	}
}

func captureOutput(f func(output *os.File)) string {
	rescueStdout := os.Stdout
	reader, writer, _ := os.Pipe()
	os.Stdout = writer

	f(writer)

	writer.Close()
	out, _ := ioutil.ReadAll(reader)
	os.Stdout = rescueStdout

	return string(out)
}


colorify_benchmark_test.go
package colorify

import (
	"os"
	"testing"
)

func BenchmarkColorize(b *testing.B) {
	colorized := NewColorable(os.Stdout)

	b.ReportAllocs()
	b.ResetTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			colorized.Orange("testString")
			colorized.Set(Style{
				Foreground: &ColorValue{
					Red:   255,
					Green: 0,
					Blue:  255,
				},
				Background: &ColorValue{
					Red:   255,
					Green: 0,
					Blue:  155,
				},
				Font: []FontEffect{
					Bold, Italic, Underline,
				},
			})
			colorized.Reset()
		}
	})
}


colorify.go
package colorify

import (
	"fmt"
	"io"
	"strconv"
	"strings"
)

type (
	// Colorable interface for console colors.
	Colorable struct {
		output io.Writer
		style  *Style
	}

	// ColorValue for the RGB value.
	ColorValue struct {
		Red   int
		Green int
		Blue  int
	}

	// Style to be applied on the text.
	Style struct {
		Foreground *ColorValue
		Background *ColorValue
		Font       []FontEffect
	}

	// FontEffect value.
	FontEffect int
)

const (
	// ResetFormat for single value, e.g. \x1b[96m
	ResetFormat = "\u001b[%dm"
	// FormatMultiValue for multiple values, e.g. \x1b[1;96m
	FormatMultiValue = "\x1b[%sm"

	foreground = 38
	background = 48
)

// Font effects.
// Some of the effects are not supported on all terminals.
const (
	Normal FontEffect = iota
	Bold
	Faint
	Italic
	Underline
	BlinkSlow
	BlinkRapid
	ReverseVideo
	Concealed
	CrossedOut
)

// NewColorable allocates and returns a new Colorable.
func NewColorable(output io.Writer) *Colorable {
	return &Colorable{
		output: output,
		style:  &Style{},
	}
}

// SetForeground color.
func (c *Colorable) SetForeground(red, green, blue int) *Colorable {
	c.style.Foreground = &ColorValue{
		Red:   red,
		Green: green,
		Blue:  blue,
	}

	return c
}

// SetBackground color.
func (c *Colorable) SetBackground(red, green, blue int) *Colorable {
	c.style.Background = &ColorValue{
		Red:   red,
		Green: green,
		Blue:  blue,
	}

	return c
}

// SetFont effects.
func (c *Colorable) SetFont(fontEffects ...FontEffect) *Colorable {
	effects := make([]FontEffect, len(fontEffects))

	for _, fontEffect := range fontEffects {
		effects = append(effects, fontEffect)
	}

	c.style.Font = effects

	return c
}

// Set a style for the next output operations.
func (c *Colorable) Set(style Style) *Colorable {
	fmt.Fprintf(c.output, c.format(style))

	return c
}

// Reset the color value to the default.
func (c *Colorable) Reset() *Colorable {
	fmt.Fprintf(c.output, resetFormat())

	return c
}

// Wrap wraps a passed a string with a color values.
func (c *Colorable) Wrap(str string, style Style) string {
	return fmt.Sprintf("%s%s%s", c.format(style), str, resetFormat())
}

func (c *Colorable) format(style Style) string {
	return fmt.Sprintf(FormatMultiValue, sequence(style))
}

func resetFormat() string {
	return fmt.Sprintf(ResetFormat, Normal)
}

func sequence(style Style) string {
	format := make([]string, 0)

	if style.Foreground != nil {
		format = append(
			format,
			fmt.Sprintf("%d;2", foreground),
			strconv.FormatInt(int64(style.Foreground.Red), 10),
			strconv.FormatInt(int64(style.Foreground.Green), 10),
			strconv.FormatInt(int64(style.Foreground.Blue), 10),
		)
	}

	if style.Background != nil {
		format = append(
			format,
			fmt.Sprintf("%d;2", background),
			strconv.FormatInt(int64(style.Background.Red), 10),
			strconv.FormatInt(int64(style.Background.Green), 10),
			strconv.FormatInt(int64(style.Background.Blue), 10),
		)
	}

	if style.Font != nil {
		for _, fontEffect := range style.Font {
			format = append(format, strconv.FormatInt(int64(fontEffect), 10))
		}
	}

	return strings.Join(format, ";")
}

// Sample of available colors.
func (c *Colorable) Sample() string {
	sample := make([]string, 0)
	for colorIndex := 0; colorIndex <= 255; colorIndex++ {
		red := (colorIndex + 5) % 255
		green := colorIndex * 3 % 255
		blue := 255 - colorIndex

		style := Style{
			Background: &ColorValue{
				Red:   red,
				Green: green,
				Blue:  blue,
			},
		}
		sample = append(sample,
			c.Wrap(
				fmt.Sprintf(
					" %-3s, %-3s, %-3s ",
					strconv.FormatInt(int64(red), 10),
					strconv.FormatInt(int64(green), 10),
					strconv.FormatInt(int64(blue), 10),
				),
				style,
			),
			" ",
		)

		if colorIndex > 0 && (colorIndex-9)%10 == 0 {
			sample = append(sample, "\n")
		}
	}

	return strings.Join(sample, "")
}

// Black color effect.
func (c *Colorable) Black(str string) string {
	return c.Wrap(str, getForegroundStyle(0, 0, 0))
}

// Blue color effect.
func (c *Colorable) Blue(str string) string {
	return c.Wrap(str, getForegroundStyle(0, 0, 255))
}

// Cyan color effect.
func (c *Colorable) Cyan(str string) string {
	return c.Wrap(str, getForegroundStyle(0, 255, 255))
}

// Gray color effect.
func (c *Colorable) Gray(str string) string {
	return c.Wrap(str, getForegroundStyle(128, 128, 128))
}

// Green color effect.
func (c *Colorable) Green(str string) string {
	return c.Wrap(str, getForegroundStyle(0, 255, 0))
}

// Magenta color effect.
func (c *Colorable) Magenta(str string) string {
	return c.Wrap(str, getForegroundStyle(255, 0, 255))
}

// Orange color effect.
func (c *Colorable) Orange(str string) string {
	return c.Wrap(str, getForegroundStyle(255, 165, 0))
}

// Red color effect.
func (c *Colorable) Red(str string) string {
	return c.Wrap(str, getForegroundStyle(255, 0, 0))
}

func getForegroundStyle(red, green, blue int) Style {
	return Style{
		Foreground: &ColorValue{
			Red:   red,
			Green: green,
			Blue:  blue,
		},
	}
}


.circleci/versionize.sh
#!/usr/bin/env bash

set -o errexit
set -o nounset
set -euo pipefail

mkdir -p $HOME/.ssh/ && echo -e "Host github.com\n\tStrictHostKeyChecking no\n" > ~/.ssh/config

LAST_COMMIT=`git log -1 --pretty=%B`
touch ./.tag
if VERSION=`git describe --abbrev=0 --tags` && [ ! -z "`git diff $VERSION`" -o -z "$VERSION" ]; then
  VERSION=${VERSION:-'0.0.0'}
  MAJOR="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  MINOR="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  PATCH="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  if echo $LAST_COMMIT | grep "\[\(major\|MAJOR\)\]" > /dev/null; then
    MAJOR=$((MAJOR+1))
    echo "$MAJOR.0.0" > ./.tag
  elif echo $LAST_COMMIT | grep "\[\(minor\|MINOR\)\]" > /dev/null; then
    MINOR=$((MINOR+1))
    echo "$MAJOR.$MINOR.0" > ./.tag
  elif echo $LAST_COMMIT | grep "\[\(patch\|PATCH\)\]" > /dev/null; then
    PATCH=$((PATCH+1))
    echo "$MAJOR.$MINOR.$PATCH" > ./.tag
  fi
else
  if echo $LAST_COMMIT | grep "\[\(major\|MAJOR\)\]" > /dev/null; then
    echo "v1.0.0" > ./.tag
  elif echo $LAST_COMMIT | grep "\[\(minor\|MINOR\)\]" > /dev/null; then
    echo "v0.1.0" > ./.tag
  elif echo $LAST_COMMIT | grep "\[\(patch\|PATCH\)\]" > /dev/null; then
    echo "v0.0.1" > ./.tag
  fi
fi


example/main.go
package main

import (
	"github.com/ahmedkamals/colorify"
	"os"
)

func main() {
	colorize := colorify.NewColorable(os.Stdout)
	style := colorify.Style{
		Foreground: &colorify.ColorValue{
			Red:   88,
			Green: 188,
			Blue:  88,
		},
		Background: &colorify.ColorValue{
			Red:   188,
			Green: 88,
			Blue:  8,
		},
		Font: []colorify.FontEffect{
			colorify.Bold,
			colorify.Italic,
			colorify.Underline,
			colorify.CrossedOut,
		},
	}

	println(colorize.Wrap("Hello styled", style))
	println(colorize.Black("Text in black!"))
	println(colorize.Blue("Deep clue C!"))
	println(colorize.Cyan("Hello cyan!"))
	println(colorize.Gray("Gray logged text!"))
	println(colorize.Green("50 shades of Green!"))
	println(colorize.Magenta("Magenta!"))
	println(colorize.Red("The thin Red light!"))
	println(colorize.Orange("Orange is the new black!"))

	colorize.Set(colorify.Style{
		Foreground: &colorify.ColorValue{
			Red:   255,
			Green: 188,
			Blue:  88,
		},
		Font: []colorify.FontEffect{colorify.Bold},
	})
	print("Output will be styled\ntill next reset!")
	colorize.Reset()
	println("\n\nSample Colors\n==============\n")
	println(colorize.Sample())
}
